### PR TITLE
target/riscv: fix addressing in `dm_read`/`dm_write`

### PR DESCRIPTION
**Pull request by en-sc**
There was an error in `dm_read`/`dm_write`: DMI address was checked against DM registers disregarding DM base address.

To solve the issue `dmi_address()` function was introduced.

Change-Id: Ia3be619417b5f5b53db5dfe302db05170d6787c9

### CODE REVIEW

File: src/target/riscv/riscv.h [hunk 1]
```diff
@@ -242,8 +242,11 @@ struct riscv_info {
 	int (*dmi_read)(struct target *target, uint32_t *value, uint32_t address);
 	int (*dmi_write)(struct target *target, uint32_t address, uint32_t value);
 
-	int (*dm_read)(struct target *target, uint32_t *value, uint32_t address);
-	int (*dm_write)(struct target *target, uint32_t address, uint32_t value);
+	/* Get the DMI address of target's DM's register.
+	 * The function should return the passed address
+	 * if the target is not assigned a DM yet.
+	 */
+	uint32_t (*dmi_address)(const struct target *target, uint32_t address);
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.h]: Optional details:

This could be called `get_dmi_address`.

The second argument could be called `dm_address`.

```suggestion
	uint32_t (*get_dmi_address)(const struct target *target, uint32_t dm_address);
```

File: src/target/riscv/riscv.c [hunk 2]
```diff
@@ -4653,28 +4645,28 @@ static const struct command_registration riscv_exec_command_handlers[] = {
 	},
 	{
 		.name = "dmi_read",
-		.handler = riscv_dmi_read,
+		.handler = handle_riscv_dmi_read,
 		.mode = COMMAND_ANY,
 		.usage = "address",
 		.help = "Perform a 32-bit DMI read at address, returning the value."
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]: Optional detail:

While making changes here, you can also improve the help messages, if you like.

```suggestion
		.help = "Read and return 32-bit value from the given address on the RISC-V DMI bus."
```

File: src/target/riscv/riscv.c [hunk 3]
```diff
@@ -4653,28 +4645,28 @@ static const struct command_registration riscv_exec_command_handlers[] = {
...
 	},
 	{
 		.name = "dmi_write",
-		.handler = riscv_dmi_write,
+		.handler = handle_riscv_dmi_write,
 		.mode = COMMAND_ANY,
 		.usage = "address value",
 		.help = "Perform a 32-bit DMI write of value at address."
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]:
```suggestion
		.help = "Write a 32-bit value to the given address on the RISC-V DMI bus."
```

File: src/target/riscv/riscv.c [hunk 4]
```diff
@@ -4653,28 +4645,28 @@ static const struct command_registration riscv_exec_command_handlers[] = {
...
 	},
 	{
 		.name = "dm_read",
-		.handler = riscv_dm_read,
+		.handler = handle_riscv_dm_read,
 		.mode = COMMAND_ANY,
 		.usage = "reg_address",
 		.help = "Perform a 32-bit read from DM register at reg_address, returning the value."
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]:
```suggestion
		.help = "Read and return 32-bit value from a debug module's register at reg_address."
```

File: src/target/riscv/riscv.c [hunk 5]
```diff
@@ -4653,28 +4645,28 @@ static const struct command_registration riscv_exec_command_handlers[] = {
...
 	},
 	{
 		.name = "dm_write",
-		.handler = riscv_dm_write,
+		.handler = handle_riscv_dm_write,
 		.mode = COMMAND_ANY,
 		.usage = "reg_address value",
 		.help = "Write a 32-bit value to the DM register at reg_address."
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]:
```suggestion
		.help = "Write a 32-bit value to the debug module's register at reg_address."
```

File: src/target/riscv/riscv-013.c [hunk 6]
```diff
@@ -691,6 +691,16 @@ static int dmi_write_exec(struct target *target, uint32_t address,
 	return dmi_op(target, NULL, NULL, DMI_OP_WRITE, address, value, true, ensure_success);
 }
 
+static uint32_t riscv_013_dmi_address(const struct target *target, uint32_t address)
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c]: If you like:

```suggestion
static uint32_t riscv_013_get_dmi_address(const struct target *target, uint32_t dm_address)
```

File: src/target/riscv/riscv.c [hunk 7]
```diff
@@ -3690,143 +3690,135 @@ COMMAND_HANDLER(riscv_authdata_write)
...
+	if (progbuf_touched || dm_deactivated) {
+		if (r->invalidate_cached_progbuf) {
+			/* Here the return value of invalidate_cached_progbuf()
+			 * is ignored. It is okay to do so for now, since the
+			 * only case an error is returned is a failure to
+			 * assign a DM to the target, which would have already
+			 * caused an error during dmi_write().
+			 * FIXME: invalidate_cached_progbuf() should be void.
+			 */
+			r->invalidate_cached_progbuf(target);
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]: Fixing of this part was the main motivation for this pull request, am I right?

Side comment: I have been considering dropping the progbuf cache completely in future - for reliability and robustness. It could be replaced by populating the program buffer by a single batch of DMI transactions, leading to high performance and no issues with maintaining the cache (and invalidating it).

File: src/target/riscv/riscv.h [hunk 8]
```diff
@@ -242,8 +242,11 @@ struct riscv_info {
 	int (*dmi_read)(struct target *target, uint32_t *value, uint32_t address);
 	int (*dmi_write)(struct target *target, uint32_t address, uint32_t value);
 
-	int (*dm_read)(struct target *target, uint32_t *value, uint32_t address);
-	int (*dm_write)(struct target *target, uint32_t address, uint32_t value);
+	/* Get the DMI address of target's DM's register.
+	 * The function should return the passed address
+	 * if the target is not assigned a DM yet.
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.h:247]: I am a bit uncertain about this part. I'd prefer to raise an error - to make sure the function cannot be used until the target is assigned a DM.

But this can be handled later (if and when).

**Review by JanMatCodasip — DISMISSED** [review]
LGTM, thank you for the fix. 

I only have few comments related to details.

File: src/target/riscv/riscv.h (hunk 8 above)
Comment (en-sc) [reply src/target/riscv/riscv.h:247]: I agree with you. This change is a bit questionable.
The reason for the current solution is `get_dm()` function. Currently `get_dm()` function:
* Returns `NULL` if a DM was not allocated/selected prior and allocation failed (out of memory).
* Returns non-examined DM (the corresponding structure is just initialized, but is yet to be filled with correct values) if none of the targets, corresponding to this DM were examined.
* Returns a properly examined DM otherwise.

The second case is quite worrying. I would like to make this function a pure accessor and search for DM / allocate it explicitly in the `examine()`. I'm preparing the corresponding patch. This would make it easier to reason about current state of a DM and allow to replace this weird behavior of `get_dmi_address()` with an assertion.

File: src/target/riscv/riscv.h (hunk 1 above)
Comment (en-sc) [reply src/target/riscv/riscv.h]: Addressed.

File: src/target/riscv/riscv.c (hunk 2 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

File: src/target/riscv/riscv.c (hunk 3 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

File: src/target/riscv/riscv.c (hunk 4 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

File: src/target/riscv/riscv.c (hunk 5 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

File: src/target/riscv/riscv-013.c (hunk 6 above)
Comment (en-sc) [reply src/target/riscv/riscv-013.c]: Addressed.

File: src/target/riscv/riscv.c (hunk 7 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Yes. You are right.

> I have been considering dropping the progbuf cache completely in future

I have conflicted feelings about your suggestion. Using batched operations will significantly speed-up writing program buffer, though still there is a benefit in implementing program buffer cache.
Consider reading a bunch of FPRs/CSRs via program buffer. The first instruction changes, but the second one is always `ebreak`. Program buffer cache would save a scan on each read even in the case of batched operations.

File: src/target/riscv/riscv.h (hunk 8 above)
Comment (JanMatCodasip) [reply src/target/riscv/riscv.h:247]: > I would like to make this function a pure accessor and search for DM / allocate it explicitly in the examine(). I'm preparing the corresponding patch.

That sounds like good solution.

File: src/target/riscv/riscv.c (hunk 7 above)
Comment (JanMatCodasip) [reply src/target/riscv/riscv.c]: >  [...] there is a benefit in implementing program buffer cache. Consider reading a bunch of FPRs/CSRs via program buffer. The first instruction changes, but the second one is always ebreak. Program buffer cache would save a scan on each read even in the case of batched operations

I assume that when the batches are used, the performance benefit would be small (and will later collect measurements to confirm).

I'd prefer robustness and reliability (even at minor performance cost). Progbuf cache in my opinion works against the reliability. 

For example:
- Program buffer may be (per the RISC-V spec) memory-mapped, and the contents of the memory can change by a program writing to it (accidentally or on purpose).
- Program buffer contents may change by direct DMI access, and our code only catches a subset of the cases.
- Program buffer contents might change by other DMI bus master accessing the registers of the debug module, without OpenOCD knowing.

In other words, I believe caches are good when they are managed reliably (flushed and invalidated as needed). When there are known gaps, I'd prefer to not use them at all.

Comment (JanMatCodasip) [reply src/target/riscv/riscv.c]: In any case, this was only a side-discussion. No need to change anything in this merge request. :-)

Marking the thread as resolved.

**Review by JanMatCodasip — APPROVED** [review]
